code/modules/reagents/chemistry/holder.dm

```diff
@@ -385,181 +689,510 @@
+
+	if(length(mix_message)) //This is only at the end
+		my_atom.audible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message.Join()]</span>")
+
+	if(!LAZYLEN(reaction_list))
+		finish_reacting()
+	else
+		update_total()
+		handle_reactions()
+
+/*
+* This ends a single instance of an ongoing reaction
+*
+* Arguments:
+* * E - the equilibrium that will be ended
+* Returns:
+* * mix_message - the associated mix message of a reaction
+*/
+/datum/reagents/proc/end_reaction(datum/equilibrium/equilibrium)
+	equilibrium.reaction.reaction_finish(src, equilibrium, equilibrium.reacted_vol)
+	if(!equilibrium.holder || !equilibrium.reaction) //Somehow I'm getting empty equilibrium. This is here to handle them
+		LAZYREMOVE(reaction_list, equilibrium)
+		qdel(equilibrium)
+		stack_trace("The equilibrium datum currently processing in this reagents datum had a nulled holder or nulled reaction. src holder:[my_atom] || src type:[my_atom.type] ") //Shouldn't happen. Does happen
+		return
+	if(equilibrium.holder != src) //When called from Destroy() eqs are nulled in smoke. This is very strange. This is probably causing it to spam smoke because of the runtime interupting the removal.
+		stack_trace("The equilibrium datum currently processing in this reagents datum had a desynced holder to the ending reaction. src holder:[my_atom] | equilibrium holder:[equilibrium.holder.my_atom] || src type:[my_atom.type] | equilibrium holder:[equilibrium.holder.my_atom.type]")
+		LAZYREMOVE(reaction_list, equilibrium)
+
+	var/reaction_message = equilibrium.reaction.mix_message
+	if(equilibrium.reaction.mix_sound)
+		playsound(get_turf(my_atom), equilibrium.reaction.mix_sound, 80, TRUE)
+	qdel(equilibrium)
 	update_total()
-	return 0
+	SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
+	return reaction_message
+
+/*
+* This stops the holder from processing at the end of a series of reactions (i.e. when all the equilibriums are completed)
+*
+* Also resets reaction variables to be null/empty/FALSE so that it can restart correctly in the future
+*/
+/datum/reagents/proc/finish_reacting()
+	STOP_PROCESSING(SSreagents, src)
+	is_reacting = FALSE
+	//Cap off values
+	for(var/_reagent in reagent_list)
+		var/datum/reagent/reagent = _reagent
+		reagent.volume = round(reagent.volume, CHEMICAL_VOLUME_ROUNDING)//To prevent runaways.
+	LAZYNULL(previous_reagent_list) //reset it to 0 - because any change will be different now.
+	update_total()
+	if(!QDELING(src))
+		handle_reactions() //Should be okay without. Each step checks.
+
+/*
+* Force stops the current holder/reagents datum from reacting
+*
+* Calls end_reaction() for each equlilbrium datum in reaction_list and finish_reacting()
+* Usually only called when a datum is transfered into a NO_REACT container
+*/
+/datum/reagents/proc/force_stop_reacting()
+	var/list/mix_message = list()
+	for(var/_equilibrium in reaction_list)
+		var/datum/equilibrium/equilibrium = _equilibrium
```

Comment:
There's more of these smelly loops but I'm sure you can find the rest on your own 
```suggestion
	for(var/datum/equilibrium/equilibrium as() in reaction_list)
```